File: ts/balanced_binary_tree.ts

class TreeNode {
  val: number
  left: TreeNode | null
  right: TreeNode | null
  constructor(val?: number, left?: TreeNode | null, right?: TreeNode | null) {
    this.val = (val === undefined ? 0 : val)
    this.left = (left === undefined ? null : left)
    this.right = (right === undefined ? null : right)
  }
}


function isBalanced(root: TreeNode | null): boolean {
  let result = true

  function dfs(root: TreeNode | null): number {
    if (!root) return 0;

    const left = dfs(root.left) as number;
    const right = dfs(root.right) as number;

    if (Math.abs(left - right) > 1) result = false

    return Math.max(left, right) + 1
  }

  dfs(root)
  return result
};

File: easy/ruby/balanced_binary_tree.rb
# Definition for a binary tree node.
# class TreeNode
#     attr_accessor :val, :left, :right
#     def initialize(val = 0, left = nil, right = nil)
#         @val = val
#         @left = left
#         @right = right
#     end
# end
# @param {TreeNode} root
# @return {Boolean}
def is_balanced(root)
  helper = lambda do |node|
    return 0 unless node

    left = helper.call(node.left)
    right = helper.call(node.right)

    return -1 if left == -1 || right == -1 || (left - right).abs > 1

    [left, right].max + 1
  end

  helper.call(root) != -1
end
